### ISSUE TITLE
bindgen header, link, etc functions take String instead of Path(Buf)

### ISSUE THREAD

**Issue by spease** [rust-lang-nursery/rust-bindgen#1297]
Probably pretty self-explanatory.

header() and link() (and other functions) on bindgen take Into<String>, even though the examples show them being used with paths (although not Path with a capital P).

I'd think this would be backwards compatible, since PathBuf is supposed to contain a superset of the characters as String.

**Comment by tangmi** [comment]
As an anecdote, I tried to build and generate bindings for a library using `cc` and `bindgen`. The shared root folder of the library is a `PathBuf`. `cc` uses `AsRef<Path>` for it's methods, such as [`cc::Build::file`](https://docs.rs/cc/1.0.1/cc/struct.Build.html#method.file) so a `PathBuf` works fine, but `bindgen` uses `Into<String>`, which means I need to call `path.into_os_string().into_string().unwrap()` or similar.

As a question for the package maintainers: does using `AsRef<Path>` make sense for `bindgen`?

**Comment by emilio** [comment]
Yes, I think it does, and I think I'd take a PR to update this for the arguments that take a path.

**Comment by tangmi** [comment]
I'm happy to take a shot at this.

Do you think the methods should just take a `impl AsRef<Path>` and immediately convert them to `String` (and fail otherwise) or if the backing data structures should store a `PathBuf`? It seems like at some point we need them in some string-ified form to pass to `clang`.

**Comment by emilio** [comment]
Hmm, yeah, that's a good point... So ideally we'd turn the clang arguments into a bunch of `Box<[u8]>` or `Vec<u8>` (so we don't need to turn it into a string) and then pass it to clang, which really wants just C strings (see `TranslationUnti::parse` in `src/clang.rs`). Though that may need `clang-sys` changes as well.

**Comment by tangmi** [comment]
I'll take a peek at it later. Some notes I've made:

- `Path` can be converted to `std::ffi::OsStr` losslessly
- There's the `std::ffi::CStr` type, which we could use instead of `Vec<u8>` to represent C strings, though I'm not sure yet how that relates to `OsStr` or to `TranslationUnite::parse`

**Comment by tangmi** [comment]
@emilio I'm struggling to find a solution that doesn't do a lossy conversion between string types and am looking for advice.

The main issue(s) I am running into are:
- `Path` and `PathBuf` are represented in a platform-specific way (UTF-8 on Unix and [WTF-8](https://simonsapin.github.io/wtf-8/) on Windows).
  - These can be converted losslessly to `OsStr` and `OsString`, respectively.
  - It seems [all `str`s are `OsStr`s](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#impl-AsRef%3COsStr%3E-2), but [not all `OsStr`s are `str`s](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.to_str)
- [`Command::arg[s]`](https://doc.rust-lang.org/std/process/struct.Command.html#method.arg) takes `AsRef<OsStr>`, which poses no problem, as we can convert `Builder::command_line_flags`, `Builder::dump_preprocessed_input`, and `Builder::rustfmt_generated_string` to build a `Vec<OsString>` instead of a `Vec<String>`.
  - I looked at [`cc::Build`](https://docs.rs/cc/1.0.28/cc/struct.Build.html), which basically wraps a `Command` that calls a compiler and builds arguments lists with `AsRef<OsStr>`.
- `bindgen` uses `clang-sys`, which takes a `*const c_char` (or `CString`).
- `CStr`/`CString` work with `Str`/`String` on all platforms, but not nessecarily with `OsStr`/`OsString` (there exists [std::os::unix::ffi::OsStrExt::as_bytes](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.as_bytes) on Unix only).

We need data as:
1. `&[u8]` for writing to files
2. `AsRef<Path>` for use in `Command::arg`
3. `*const c_char`/`CStr` for use in `clang_sys`

From this, it seems that on Unix, we can use `AsRef<Path>` from user input and convert to the proper string types. On Windows, however, there has to be some lossy conversion (or error propagation) at some point, as we can't guarantee that conversions for 1. and 3. can be done safely.

I'd like to represent `Builder` and `BindgenOptions` with `PathBuf` for the relevant fields, as it seems more semantic than `String` (or alternatives), but before I go too far down that path, I'd like to know your thoughts on this.

Thanks!

**Comment by pvdrz** [comment]
I see three paths here:
1. Handle windows via the lossy path: `OsString` -> `String` -> `CString` and return an error when trying to set a clang_arg with an `OsString` that cannot be converted in this way. 
2. Change the unix methods only so they take `AsRef<OsStr>` instead and let windows keep using `AsRef<str>` and `Into<String>`.
3. Leave things as they are right now.

**Comment by dario23** [comment]
i'm not a windows developer either, but i'd like to have a go at this, and would try option 1) first.
